nixos/modules/services/audio/alsa.nix

```diff
@@ -1,38 +1,439 @@
+              }
+              capture.pcm $A
+            }
 
+            # Device which records and plays back audio
+            pcm.recorder {
+              type asym
+              capture.pcm {
+                type dsnoop
+                ipc_key 9165218
+                ipc_perm 0666
+                slave.pcm "hw:loopback,1,0"
+                slave.period_size 1024
+                slave.buffer_size 8192
+              }
+              playback.pcm {
+                type dmix
+                ipc_key 6181923
+                ipc_perm 0666
+                slave.pcm "hw:loopback,0,0"
+                slave.period_size 1024
+                slave.buffer_size 8192
+              }
+            }
+          '')
+          (lib.mapAttrsToList mkControl cfg.controls)
+          (lib.mapAttrsToList (n: v: "pcm.${n} ${quote v}") cfg.deviceAliases)
+        ];
+      in
+      lib.mkBefore (lib.concatStringsSep "\n" (lib.flatten conf));
+
+    hardware.alsa.cardAliases = lib.mkIf cfg.enableRecorder {
+      loopback.driver = "snd_aloop";
+      loopback.id = 2;
+    };
+
+    # Set default PCM devices
+    environment.sessionVariables = defaultDeviceVars;
+    systemd.globalEnvironment = defaultDeviceVars;
+
+    environment.etc."asound.conf".text = cfg.config;
+
+    boot.kernelModules =
+      [ ]
+      ++ lib.optionals cfg.enableOSSEmulation [ "snd_pcm_oss" "snd_mixer_oss" ]
+      ++ lib.optionals cfg.enableRecorder [ "snd_aloop" ];
+
+    # Assign names to the sound cards
+    boot.extraModprobeConfig = lib.concatStringsSep "\n" cardsConfig;
+
+    # Provide alsamixer, aplay, arecord, etc.
+    environment.systemPackages = [ pkgs.alsa-utils ];
+
+    # Install udev rules for restoring card settings on boot
+    services.udev.extraRules = ''
+      ACTION=="add", SUBSYSTEM=="sound", KERNEL=="controlC*", KERNELS!="card*", GOTO="alsa_restore_go"
+      GOTO="alsa_restore_end"
+
+      LABEL="alsa_restore_go"
+      TEST!="/etc/alsa/state-daemon.conf", RUN+="${alsactl} restore -gU $attr{device/number}"
+      TEST=="/etc/alsa/state-daemon.conf", RUN+="${alsactl} nrestore -gU $attr{device/number}"
+      LABEL="alsa_restore_end"
+    '';
+
+    # Service to store/restore the sound card settings
     systemd.services.alsa-store = {
       description = "Store Sound Card State";
       wantedBy = [ "multi-user.target" ];
+      restartIfChanged = false;
       unitConfig = {
         RequiresMountsFor = "/var/lib/alsa";
         ConditionVirtualization = "!systemd-nspawn";
       };
       serviceConfig = {
         Type = "oneshot";
         RemainAfterExit = true;
+        # Note: the service should never be restated, otherwise any
+        # setting changed between the last `store` and now will be lost.
+        # To prevent NixOS from starting it in case it has failed we
+        # expand the exit codes considered successful
+        SuccessExitStatus = [ 0 99 ];
         ExecStartPre = "${pkgs.coreutils}/bin/mkdir -p /var/lib/alsa";
```

Comment:
This should be using tmpfiles.

Reply:
How do you order the service after tmpfiles having created the directory?

Reply:
Used `StateDirectory=` instead.